src/zero_copy/error.rs

```diff
@@ -1,9 +1,33 @@
 use super::*;
 
-pub trait Error<T> {
-    fn create() -> Self;
+pub trait Error<I>: Sized
+where
+    I: ?Sized + Input,
+{
+    type Label;
+
+    fn expected_input_found<Iter: IntoIterator<Item = Option<I::Token>>>(span: I::Span, expected: Iter, found: Option<I::Token>) -> Self {
+        Self::custom("expected input found", span, expected.into_iter().chain([found]))
+    }
+
+    fn unclosed_delimiter(unclosed_span: I::Span, unclosed: I::Token, span: I::Span, expected: I::Token, found: Option<I::Token>) -> Self {
+        #![allow(unused_variables)]
+        Self::custom("unclosed delimiter", span, [Some(unclosed), Some(expected), found])
+    }
+
+    fn custom<Iter: IntoIterator<Item = Option<I::Token>>>(message: &str, span: I::Span, context: Iter) -> Self;
```

Comment:
I know you mentioned wanting to have some API for more 'user-defined' errors, and this was my first go at it. I feel like it might be nice to have some more standard way to have at the least an error with Span and Message on the trait, for user-defined generic parsers (I know I wished it for it once while working on a library for binary parsers).

Reply:
Perhaps this should be an extension trait then? `trait CustomError<I>: Error<I>`, say. It might be worth putting these sort of utilities behind a feature flag though. We could do this in another PR though.

Reply:
Sounds good to me, I'll switch it over